waspc/src/Analyzer/TypeChecker/Internal.hs

```diff
@@ -0,0 +1,232 @@
+
+checkAST :: AST -> T TypedAST
+checkAST (P.AST stmts) = TypedAST <$> mapM checkStmt stmts
+
+checkStmt :: P.Stmt -> T TypedStmt
+checkStmt (P.Decl typName name expr) =
+  lookupDeclType typName >>= \case
+    Nothing -> throw $ NoDeclarationType typName
+    Just (TD.DeclType _ expectedType) -> do
+      -- Decides whether the argument to the declaration has the correct type
+      mTypedExpr <- weaken expectedType <$> checkExpr expr
+      case mTypedExpr of
+        Left e -> throw e
+        Right typedExpr -> return $ Decl name typedExpr (DeclType typName)
+
+-- | Determine the type of an expression, following the inference rules described in
+-- the wasplang document. Some these rules are referenced by name in the comments
+-- of the following functions using [Brackets].
+checkExpr :: P.Expr -> T TypedExpr
+checkExpr (P.StringLiteral s) = return $ StringLiteral s
+checkExpr (P.IntegerLiteral i) = return $ IntegerLiteral i
+checkExpr (P.DoubleLiteral d) = return $ DoubleLiteral d
+checkExpr (P.BoolLiteral b) = return $ BoolLiteral b
+checkExpr (P.ExtImport n s) = return $ ExtImport n s
+checkExpr (P.Identifier ident) =
+  lookupType ident >>= \case
+    Nothing -> throw $ UndefinedIdentifier ident
+    Just typ -> return $ Var ident typ
+-- For now, the two quoter types are hardcoded here, it is an error to use a different one
+-- TODO: this will change when quoters are added to "Analyzer.TypeDefinitions"
+checkExpr (P.Quoter "json" s) = return $ JSON s
+checkExpr (P.Quoter "psl" s) = return $ PSL s
+checkExpr (P.Quoter tag _) = throw $ QuoterUnknownTag tag
+-- The type of a list is the unified type of its values.
+-- This poses a problem for empty lists, there is not enough information to choose a type.
+-- TODO: fix this in the future, probably by adding an additional phase to resolve type variables
+--       that would be assigned here
+checkExpr (P.List values) = do
+  typedValues <- mapM checkExpr values
+  case unify <$> nonEmpty typedValues of
+    Nothing -> throw EmptyListNotImplemented
+    Just (Left e) ->
+      throw e
+    Just (Right (unifiedValues, unifiedType)) ->
+      return $ List (toList unifiedValues) (ListType unifiedType)
+-- Apply [Dict], and also check that there are no duplicate keys in the dictionary
+checkExpr (P.Dict entries) = do
+  guardUnique $ map fst entries
+  typedEntries <- zip (map fst entries) <$> mapM (checkExpr . snd) entries
+  let dictType = M.fromList $ map (\(key, val) -> (key, DictRequired $ exprType val)) typedEntries
+  return $ Dict typedEntries (DictType dictType)
+  where
+    guardUnique :: [String] -> T ()
+    guardUnique [] = pure ()
+    guardUnique (x : xs)
+      | x `notElem` xs = guardUnique xs
+      | otherwise = throw $ DictDuplicateField x
+
+-- | @unify exprs == Right (exprs', typ)@ shows that all expressions in @exprs@
+-- can be typed as @typ@. Additionally, The following property is guaranteed:
+--
+-- * @all ((==typ) . exprType) exprs == True@
```

Comment:
All right, that makes sense then!
Would you mind adding verbal explanation to this formula, exactly as you described it above: "it's just saying that unify changes the type stored in each TypedExpr to be the same as the type it unifies to"? It will make it quicker to understand + harder to introduce typos in the future as it contains more redundancy.